harry-core/src/harry/model/sut/injvm/InJVMTokenAwareVisitExecutor.java

```diff
@@ -72,44 +78,43 @@ public InJVMTokenAwareVisitExecutor(Run run,
+        int retries = 0;
 
-        Object[] pk =  schema.inflatePartitionKey(pd);
+        Object[] pk = schema.inflatePartitionKey(pd);
         List<TokenPlacementModel.Node> replicas = getRing().replicasFor(TokenUtil.token(ByteUtils.compose(ByteUtils.objectsToBytes(pk))));
-
-        TokenPlacementModel.Node replica = replicas.get((int) (lts % replicas.size()));
-        if (cl == SystemUnderTest.ConsistencyLevel.NODE_LOCAL)
-        {
-            future.complete(executeNodeLocal(statement.cql(), replica, statement.bindings()));
-        }
-        else
+        while (retries++ < MAX_RETRIES)
         {
-            CompletableFuture.supplyAsync(() ->  sut.cluster
-                                                 .stream()
-                                                 .filter((n) -> n.config().broadcastAddress().toString().contains(replica.id))
-                                                 .findFirst()
-                                                 .get()
-                                                 .coordinator()
-                                                 .execute(statement.cql(), InJvmSut.toApiCl(cl), statement.bindings()), executor)
-                             .whenComplete((res, t) ->
-                                           {
-                                               if (t != null)
-                                                   executor.schedule(() -> executeAsyncWithRetries(lts, pd, future, statement, retries + 1), 1, TimeUnit.SECONDS);
-                                               else
-                                                   future.complete(res);
-                                           });
+            try
+            {
+                TokenPlacementModel.Node replica = replicas.get((int) (lts % replicas.size()));
+                if (cl == SystemUnderTest.ConsistencyLevel.NODE_LOCAL)
+                {
+                    return executeNodeLocal(statement.cql(), replica, statement.bindings());
+                }
+                else
+                {
+                    return sut.cluster
+                           .stream()
+                           .filter((n) -> n.config().broadcastAddress().toString().contains(replica.id))
+                           .findFirst()
+                           .get()
+                           .coordinator()
+                           .execute(statement.cql(), InJvmSut.toApiCl(cl), statement.bindings());
+                }
+            }
+            catch (Throwable t)
+            {
+                int delaySecs = 1;
+                logger.error(String.format("Caught message while trying to execute: %s. Scheduled to retry in %s seconds", statement, delaySecs), t);
+                Uninterruptibles.sleepUninterruptibly(delaySecs, TimeUnit.SECONDS);
+            }
         }
+        throw new IllegalStateException(String.format("Can not execute statement %s after %d retries", statement, retries));
```

Comment:
So tl;dr we made this synchronous and added some logging for visibility? I'll figure out why I suppose as I look at its callers...

Reply:
Let's say this was the simplest way to know that a query _has_ in fact finished. This is a barebone model, and probably should be improved upon. But at least it gives you some ground to stand upon.